// react-native-enhanced-popup-menu/dist/MenuItem.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuItem = void 0;
const React = __importStar(require("react"));
const react_native_1 = require("react-native");
exports.MenuItem = (_a) => {
    var { disabled = false, disabledTextColor = '#BDBDBD', underlayColor = '#E0E0E0', style, textStyle, children, onPress } = _a, props = __rest(_a, ["disabled", "disabledTextColor", "underlayColor", "style", "textStyle", "children", "onPress"]);
    return (<react_native_1.TouchableHighlight disabled={disabled} onPress={onPress} style={[styles.container, style]} underlayColor={underlayColor} {...props}>
      <react_native_1.Text ellipsizeMode={react_native_1.Platform.OS === 'ios' ? 'clip' : 'tail'} numberOfLines={1} style={[
        styles.title,
        disabled && { color: disabledTextColor },
        textStyle
    ]}>
        {children}
      </react_native_1.Text>
    </react_native_1.TouchableHighlight>);
};
const styles = react_native_1.StyleSheet.create({
    container: {
        height: 48,
        justifyContent: 'center',
        maxWidth: 248,
        minWidth: 124
    },
    title: {
        fontSize: 14,
        fontWeight: '400',
        paddingHorizontal: 16
    }
});
