// react-native-calendars/src/calendar/header/style.js
import {StyleSheet} from 'react-native';
import * as defaultStyle from '../../style';
import Fonts from "UIProps/Fonts";
import {Colors} from "UIProps/Colors";
const STYLESHEET_ID = 'stylesheet.calendar.header';

export default function(theme={}) {
  const appStyle = {...defaultStyle, ...theme};
  return StyleSheet.create({
    header: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      // paddingLeft: 10,
      
      paddingRight: 10,
      marginTop: 6,
      alignItems: 'center'
    },
    headerContainer: {
      flexDirection: 'row',
      
    },
    monthText: {
      fontSize: appStyle.textMonthFontSize,
      fontFamily:Fonts.medium,
      // fontWeight: appStyle.textMonthFontWeight,
      color: '#9E9E9E',
      margin: 10
    },
    arrow: {
      padding: 10,
      paddingHorizontal:2,
      ...appStyle.arrowStyle
    },
    arrowImage: {
      tintColor: Colors.accent
    },
    disabledArrowImage: {
      tintColor: appStyle.disabledArrowColor
    },
    week: {
      marginTop: 7,
      flexDirection: 'row',
      justifyContent: 'space-around'
    },
    dayHeader: {
      marginTop: 2,
      marginBottom: 7,
      width: 32,
      textAlign: 'center',
      fontSize: appStyle.textDayHeaderFontSize,
      fontFamily: appStyle.textDayHeaderFontFamily,
      fontWeight: appStyle.textDayHeaderFontWeight,
      color: appStyle.textSectionTitleColor
    },
    disabledDayHeader: {
      color: appStyle.textSectionTitleDisabledColor
    },
    ...(theme[STYLESHEET_ID] || {})
  });
}
